uber/models.py

```diff
@@ -494,93 +494,138 @@ def lookup_attendee(self, full_name, email, zip_code):
-                                                 Attendee.badge_num != 0):
-                a.badge_num += shift
+            return new_badge_num
 
-        def change_badge(self, attendee, badge_type, badge_num=None):
+        def update_badge(self, attendee, old_badge_type, old_badge_num):
             """
-            Badges should always be assigned a number if they're marked as
-            pre-assigned or if they've been checked in.  If auto-shifting is
-            also turned off, badge numbers cannot clobber other numbers,
-            otherwise we'll shift all the other badge numbers around the old
-            and new numbers.
+            This should be called whenever an attendee's badge type or badge number is changed by an admin. It checks
+            if the attendee will still require a badge number with their new badge type, and if so, sets their number
+            to either the number specified by the admin or the lowest available badge number in that range.
+
+            :param attendee: The Attendee() object whose badge is being changed.
+            :param old_badge_type: The old badge type.
+            :param old_badge_num: The old badge number.
+            :return:
             """
-            # assert_badge_locked()
-            from uber.badge_funcs import check_range
-            badge_type = int(badge_type)
-            old_badge_type, old_badge_num = attendee.badge_type, attendee.badge_num
-
-            out_of_range = check_range(badge_num, badge_type)
-            next = self.next_badge_num(badge_type, old_badge_num)
-            if out_of_range:
-                return out_of_range
-            elif not badge_num and next > c.BADGE_RANGES[badge_type][1]:
-                return 'There are no more badges available for that type'
-            elif badge_type in c.PREASSIGNED_BADGE_TYPES and not c.SHIFT_CUSTOM_BADGES:
-                return 'Custom badges have already been ordered'
-
-            if not c.SHIFT_CUSTOM_BADGES:
-                badge_num = badge_num or next
-                if badge_num != 0:
-                    existing = self.query(Attendee).filter_by(badge_type=badge_type, badge_num=badge_num) \
-                                                   .filter(Attendee.id != attendee.id)
-                    if existing.count():
-                        return 'That badge number already belongs to {!r}'.format(existing.first().full_name)
-            else:
+            old_badge_num = int(old_badge_num or 0) or None
+
+            if old_badge_type == attendee.badge_type and (not attendee.badge_num or old_badge_num == attendee.badge_num):
+                attendee.badge_num = old_badge_num
+                return 'Attendee is already {} with badge {}'.format(c.BADGES[old_badge_type], old_badge_num)
+
+            if c.SHIFT_CUSTOM_BADGES:
                 # fill in the gap from the old number, if applicable
+                badge_num_keep = attendee.badge_num
```

Comment:
Hmm, it's not obvious at a glance why that would happen.  Are the unit tests failing for the cases where the badge type changes, where the badge type stays the same, or both?